dingo/experimental_gurobi_functions/redundant_facet_removal_experiment.py

```diff
@@ -0,0 +1,300 @@
+        new_objective_coeffs (list): List of new objective coefficients for the variables.
+
+    Returns:
+        gurobipy.Model: The updated Gurobi model with the new objective function.
+    """
+    # Clear the existing objective function
+    model.setObjective(0, clear=True)
+
+    # Update the objective function with the new coefficients
+    for i, var in enumerate(model.getVars()):
+        var.setAttr(GRB.Attr.Obj, new_objective_coeffs[i])
+
+    # Optimize the updated model
+    model.optimize()
+
+    return model
+
+def fast_remove_redundant_facets(lb, ub, S, c, opt_percentage=100):
+    if lb.size != S.shape[1] or ub.size != S.shape[1]:
+        raise Exception(
+            "The number of reactions must be equal to the number of given flux bounds."
+        )
+
+    redundant_facet_tol = 1e-07
+    tol = 1e-06
+
+    m = S.shape[0]
+    n = S.shape[1]
+    beq = np.zeros(m)
+    Aeq_res = S
+
+    A = np.zeros((2 * n, n), dtype="float")
+    A[0:n] = np.eye(n)
+    A[n:] -= np.eye(n, n, dtype="float")
+
+    b = np.concatenate((ub, -lb), axis=0)
+    b = np.asarray(b, dtype="float")
+    b = np.ascontiguousarray(b, dtype="float")
+
+    max_biomass_flux_vector, max_biomass_objective = fast_fba(lb, ub, S, c)
+    val = -np.floor(max_biomass_objective / tol) * tol * opt_percentage / 100
+
+    b_res = []
+    A_res = np.empty((0, n), float)
+    beq_res = np.array(beq)
+
+    try:
+        with gp.Env(empty=True) as env:
+            env.setParam("OutputFlag", 0)
+            env.start()
+
+            with gp.Model(env=env) as model:
+                x = model.addMVar(
+                    shape=n,
+                    vtype=GRB.CONTINUOUS,
+                    name="x",
+                    lb=lb,
+                    ub=ub,
+                )
+
+                Aeq_sparse = sp.csr_matrix(S)
+                A_sparse = sp.csr_matrix(np.array(-c))
+                b_sparse = np.array(val)
+
+                b = np.array(b)
+                beq = np.array(beq)
+
+                model.addMConstr(Aeq_sparse, x, "=", beq, name="c")
+                model.update()
+                model.addMConstr(A_sparse, x, "<", [val], name="d")
+                model.update()
+
+                model_iter = model.copy()
+
+                indices_iter = range(n)
+                removed = 1
+                offset = 1
+                facet_left_removed = np.zeros((1, n), dtype=bool)
+                facet_right_removed = np.zeros((1, n), dtype=bool)
+
+                while removed > 0 or offset > 0:
+                    removed = 0
+                    offset = 0
+                    indices = indices_iter
+                    indices_iter = []
+
+                    Aeq_sparse = sp.csr_matrix(Aeq_res)
+                    beq = np.array(beq_res)
+
+                    b_res = []
+                    A_res = np.empty((0, n), float)
+                    A_res = np.empty((0, n), float)
```

Comment:
This is written twice. If there is no reason for this, plz delete one of those two

Reply:
ok doing it